packages/vite/src/node/plugins/resolve.ts

```diff
@@ -267,6 +256,35 @@ function tryFsResolve(
   }
 }
 
+function tryFsResolve(
+  fsPath: string,
+  options: InternalResolveOptions,
+  tryIndex = true
+): string | undefined {
+  let file = fsPath
+  let postfix = ''
+  let res: string | undefined
+
+  // 1、Try resolving without removing postfixes first
+  if ((res = tryResolve(file, postfix, options, tryIndex))) {
```

Comment:
We are going to lose the optimization introduced in https://github.com/vitejs/vite/pull/2718 with this PR.

When people use imports with extensions, we should first check for an exact match before trying the list of auto-resolved extensions. Also, I think that we should first with the extensions removed to avoid two checks per file for common cases.

The order of the `tryResolveFile` calls should be:
```
  file ( tryIndex: false, removing postfixes )
  file ( tryIndex: false )

  for each ext
    file+ext ( removing postfixes )
    file+ext 

  file ( tryIndex: true, removing postfixes )
  file ( tryIndex: true )
```
I also think that if `?` and `#` are inside the folder path and are not part of the file, we can safely remove the postfixes so we avoid the double checks. So we should check for the presence of `?` and `#`  only after the last `/`. This change alone should fix most issues (the other changes in this PR are still important though).